File: diff-adapter/src/main/java/ru/starksoft/differ/adapter/DifferRefreshingData.java
package ru.starksoft.differ.adapter;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.core.util.Pools;
import ru.starksoft.differ.adapter.viewmodel.ViewModel;

/**
 * Связка списка ViewModel'ей и меток обновления
 * <p>
 * Метки передаеются в refreshAdapter(метка) и обрабатываюся в completedUpdateAdapter()
 */
class DifferRefreshingData {

	private static final int MAX_SIZE_POOL = 20;
	private static final Pools.Pool<DifferRefreshingData> refreshingDataPool = new Pools.SynchronizedPool<>(MAX_SIZE_POOL);
	@NonNull private final List<ViewModel> items = new ArrayList<>();
	@NonNull private DifferLabels labels;

	private DifferRefreshingData() {
		labels = new DifferLabels();
	}

	@NonNull
	public static DifferRefreshingData obtain() {
		DifferRefreshingData differRefreshingData = refreshingDataPool.acquire();
		if (differRefreshingData == null) {
			differRefreshingData = new DifferRefreshingData();
		}
		return differRefreshingData;
	}

	void release() {
		items.clear();
		labels.release();
		refreshingDataPool.release(this);
	}

	void init(@NonNull List<ViewModel> items, @NonNull DifferLabels labels) {
		this.items.clear();
		this.items.addAll(items);

		this.labels = new DifferLabels();
		this.labels.add(labels.getItems());
	}

	@NonNull
	List<ViewModel> getItems() {
		return items;
	}

	@NonNull
	DifferLabels getLabels() {
		return labels;
	}
}

File: diff-adapter/src/main/java/ru/starksoft/differ/adapter/DifferRefreshingData.kt
package ru.starksoft.differ.adapter

import androidx.core.util.Pools
import androidx.core.util.Pools.SynchronizedPool
import ru.starksoft.differ.adapter.viewmodel.ViewModel
import java.util.*

/**
 * Связка списка ViewModel'ей и меток обновления
 *
 *
 * Метки передаеются в refreshAdapter(метка) и обрабатываюся в completedUpdateAdapter()
 */
internal class DifferRefreshingData private constructor() {

    private val items: MutableList<ViewModel> = ArrayList()
    var labels: DifferLabels
        private set

    init {
        labels = DifferLabels()
    }

    fun release() {
        items.clear()
        labels.release()
        refreshingDataPool.release(this)
    }

    fun init(items: List<ViewModel>, labels: DifferLabels) {
        this.items.clear()
        this.items.addAll(items)
        this.labels = DifferLabels()
        this.labels.add(labels.getItems())
    }

    fun getItems(): List<ViewModel> {
        return items
    }

    companion object {

        private const val MAX_SIZE_POOL = 20
        private val refreshingDataPool: Pools.Pool<DifferRefreshingData> =
            SynchronizedPool(MAX_SIZE_POOL)

        fun obtain(): DifferRefreshingData {
            var differRefreshingData = refreshingDataPool.acquire()
            if (differRefreshingData == null) {
                differRefreshingData = DifferRefreshingData()
            }
            return differRefreshingData
        }
    }
}
